src/code/audio/audio_thread.c

```diff
@@ -0,0 +1,979 @@
+void AudioThread_Noop1CmdZeroed(void) {
+    AUDIOCMD_GLOBAL_NOOP_1(0, 0, 0, 0);
+}
+
+/**
+ * Call an audio-thread command that has no code to process it. Unused.
+ */
+void AudioThread_Noop2Cmd(u32 arg0, s32 arg1) {
+    AUDIOCMD_GLOBAL_NOOP_2(0, 0, arg1, arg0);
+}
+
+// Unused
+void AudioThread_WaitForAudioTask(void) {
+    osRecvMesg(gAudioCtx.taskStartQueueP, NULL, OS_MESG_NOBLOCK);
+    osRecvMesg(gAudioCtx.taskStartQueueP, NULL, OS_MESG_BLOCK);
+}
+
+/**
+ * Get the number of s16-samples from the start of the sample to the current sample position.
+ * Unused
+ */
+s32 AudioThread_GetSamplePosFromStart(s32 seqPlayerIndex, s32 channelIndex, s32 layerIndex) {
+    s32 pad;
+    s32 loopEnd;
+    s32 samplePosInt;
+
+    if (!AudioThread_GetSamplePos(seqPlayerIndex, channelIndex, layerIndex, &loopEnd, &samplePosInt)) {
+        return 0;
+    }
+    return samplePosInt;
+}
+
+/**
+ * Get the number of s16-samples from the current sample position to the end of the sample.
+ * Unused
+ */
+s32 AudioThread_GetSamplePosUntilEnd(s32 seqPlayerIndex, s32 channelIndex, s32 layerIndex) {
+    s32 pad;
+    s32 loopEnd;
+    s32 samplePosInt;
+
+    if (!AudioThread_GetSamplePos(seqPlayerIndex, channelIndex, layerIndex, &loopEnd, &samplePosInt)) {
+        return 0;
+    }
+    return loopEnd - samplePosInt;
+}
+
+// Only used in unused functions
+s32 AudioThread_GetSamplePos(s32 seqPlayerIndex, s32 channelIndex, s32 layerIndex, s32* loopEnd, s32* samplePosInt) {
+    SequencePlayer* seqPlayer = &gAudioCtx.seqPlayers[seqPlayerIndex];
+    SequenceLayer* layer;
+    Note* note;
+    TunedSample* tunedSample;
+
+    if (seqPlayer->enabled && seqPlayer->channels[channelIndex]->enabled) {
+        layer = seqPlayer->channels[channelIndex]->layers[layerIndex];
+        if (layer == NULL) {
+            return false;
+        }
+
+        if (layer->enabled) {
+            if (layer->note == NULL) {
+                return false;
+            }
+
+            if (!layer->bit3) {
+                return false;
+            }
+
+            note = layer->note;
+            if (layer == note->playbackState.parentLayer) {
+
+                if (note->sampleState.bitField1.isSyntheticWave == true) {
+                    return false;
+                }
+
+                tunedSample = note->sampleState.tunedSample;
+                if (tunedSample == NULL) {
+                    return false;
+                }
+                *loopEnd = tunedSample->sample->loop->loopEnd;
+                *samplePosInt = note->synthesisState.samplePosInt;
+                return true;
+            }
+            return false;
+        }
+    }
+    return false;
+}
+
+s32 AudioThread_GetEnabledNotesCount(void) {
+    return AudioThread_CountAndReleaseNotes(0);
+}
+
+s32 AudioThread_GetEnabledSampledNotesCount(void) {
+    return AudioThread_CountAndReleaseNotes(2);
+}
+
+/**
+ * @param flags 0: count notes. 1: release all notes. 2: count sample notes 3: release sample notes
```

Comment:
I'm not convinced about `flags` for this given that it seems to act more like an enum?

Reply:
It kinda is flags based though, because by default, a count of the notes is obtained. But toggling bit 1 will trigger the notes to release, and toggling bit 2 will cause only notes with samples attached to them to be processed. I'll look into making those defines